<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>Cas2Consed.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">92_jcvi-javacommon</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.assembly.clc.cas.consed</a> &gt; <span class="el_source">Cas2Consed.java</span></div><h1>Cas2Consed.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Copyright (c) 2013 J. Craig Venter Institute.
 * 	This file is part of Jillion
 * 
 * 	 Jillion is free software: you can redistribute it and/or modify
 * 	it under the terms of the GNU General Public License as published by
 * 	the Free Software Foundation, either version 3 of the License, or
 * 	(at your option) any later version.
 * 	
 * 	 Jillion is distributed in the hope that it will be useful,
 * 	but WITHOUT ANY WARRANTY; without even the implied warranty of
 * 	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * 	GNU General Public License for more details.
 * 	
 * 	You should have received a copy of the GNU General Public License
 * 	along with  Jillion.  If not, see http://www.gnu.org/licenses
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.assembly.clc.cas.consed;

import java.io.File;
import java.io.IOException;
import java.util.Date;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

import org.jcvi.jillion.assembly.clc.cas.AbstractAlignedReadCasVisitor;
import org.jcvi.jillion.assembly.clc.cas.CasGappedReferenceDataStore;
import org.jcvi.jillion.assembly.clc.cas.read.CasPlacedRead;
import org.jcvi.jillion.assembly.consed.ConsedUtil;
import org.jcvi.jillion.assembly.consed.ace.AceContig;
import org.jcvi.jillion.assembly.consed.ace.AceContigBuilder;
import org.jcvi.jillion.assembly.consed.ace.AceFileWriter;
import org.jcvi.jillion.assembly.consed.ace.AceFileWriterBuilder;
import org.jcvi.jillion.assembly.consed.ace.PhdInfo;
import org.jcvi.jillion.assembly.consed.ace.WholeAssemblyAceTag;
import org.jcvi.jillion.assembly.consed.phd.Phd;
import org.jcvi.jillion.assembly.consed.phd.PhdBallWriter;
import org.jcvi.jillion.assembly.consed.phd.PhdDataStore;
import org.jcvi.jillion.assembly.consed.phd.PhdFileDataStoreBuilder;
import org.jcvi.jillion.assembly.consed.phd.PhdWriter;
import org.jcvi.jillion.assembly.util.consensus.MostFrequentBasecallConsensusCaller;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.datastore.DataStoreException;
import org.jcvi.jillion.core.datastore.DataStoreProviderHint;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.qual.PhredQuality;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.util.DateUtil;
import org.jcvi.jillion.core.util.iter.StreamingIterator;
import org.jcvi.jillion.fasta.nt.NucleotideFastaDataStore;
import org.jcvi.jillion.fasta.nt.NucleotideFastaFileDataStoreBuilder;
import org.jcvi.jillion.trace.Trace;
import org.jcvi.jillion.trace.fastq.FastqDataStore;
import org.jcvi.jillion.trace.fastq.FastqFileDataStoreBuilder;
import org.jcvi.jillion.trace.sff.SffFileIterator;

public class Cas2Consed extends  AbstractAlignedReadCasVisitor{

	private final Map&lt;String, AceContigBuilder&gt; contigBuilders;
	private final File consedOutputDir;
<span class="nc" id="L66">	private final Date phdDate = new Date();</span>
	
	private final PhdWriter phdOut;
	
<span class="nc" id="L70">	private File chromatDir = null;</span>
	private final File phdFile;
	
	private String prefix;
	
	public Cas2Consed(File casFile,	CasGappedReferenceDataStore gappedReferenceDataStore, File consedOutputDir,
			String prefix) throws DataStoreException, IOException {
<span class="nc" id="L77">		super(casFile, gappedReferenceDataStore);</span>
		
<span class="nc bnc" id="L79" title="All 2 branches missed.">		if(consedOutputDir ==null){</span>
<span class="nc" id="L80">			throw new NullPointerException(&quot;output dir can not be null&quot;);</span>
		}
<span class="nc bnc" id="L82" title="All 2 branches missed.">		if(prefix == null){</span>
<span class="nc" id="L83">			throw new NullPointerException(&quot;prefix can not be null&quot;);</span>
		}
<span class="nc" id="L85">		this.prefix = prefix.trim();</span>
<span class="nc bnc" id="L86" title="All 2 branches missed.">		if(this.prefix.isEmpty()){</span>
<span class="nc" id="L87">			throw new IllegalArgumentException(&quot;prefix must contain non-whitespace&quot;);</span>
		}
<span class="nc" id="L89">		contigBuilders = new LinkedHashMap&lt;String, AceContigBuilder&gt;();</span>
<span class="nc" id="L90">		this.consedOutputDir = consedOutputDir;</span>
		
<span class="nc" id="L92">		StreamingIterator&lt;String&gt; referenceIdIterator = null;</span>
<span class="nc" id="L93">		StreamingIterator&lt;NucleotideSequence&gt; referenceSequenceIterator = null;</span>
		
<span class="nc" id="L95">		referenceIdIterator = gappedReferenceDataStore.idIterator();</span>
<span class="nc" id="L96">		referenceSequenceIterator = gappedReferenceDataStore.iterator();</span>
		
		try{
<span class="nc bnc" id="L99" title="All 2 branches missed.">			while(referenceIdIterator.hasNext()){</span>
<span class="nc" id="L100">				String id = referenceIdIterator.next();</span>
<span class="nc" id="L101">				NucleotideSequence seq = referenceSequenceIterator.next();</span>
<span class="nc" id="L102">				contigBuilders.put(id, new AceContigBuilder(id, seq));</span>
<span class="nc" id="L103">			}</span>
		}finally{
<span class="nc" id="L105">			IOUtil.closeAndIgnoreErrors(referenceIdIterator, referenceSequenceIterator);</span>
		}
<span class="nc" id="L107">		File phdDir = new File(consedOutputDir, &quot;phdball_dir&quot;);</span>
		
<span class="nc" id="L109">		phdFile = new File(phdDir, &quot;phd.ball.1&quot;);</span>
<span class="nc" id="L110">		IOUtil.mkdirs(phdDir);</span>
<span class="nc" id="L111">		phdOut = new PhdBallWriter(phdFile);</span>
<span class="nc" id="L112">	}</span>

	public final File getPhdBallFile() {
<span class="nc" id="L115">		return phdFile;</span>
	}

	public final void setChromatDir(File chromatDir) {
<span class="nc" id="L119">		this.chromatDir = chromatDir;</span>
<span class="nc" id="L120">	}</span>

	@Override
	protected void visitUnMatched(Trace currentTrace) {
		//no-op
		//we don't care about reads that don't align
<span class="nc" id="L126">	}</span>

	@Override
	protected final void visitMatch(String referenceId, CasPlacedRead read,
			Trace traceOfRead) {
<span class="nc" id="L131">		AceContigBuilder builder = contigBuilders.get(referenceId);</span>
<span class="nc bnc" id="L132" title="All 2 branches missed.">		if(!(traceOfRead instanceof PhdReadRecord)){</span>
<span class="nc" id="L133">			throw new IllegalStateException(&quot;not a valid phd record &quot; + traceOfRead);</span>
		}
<span class="nc" id="L135">		PhdReadRecord phdReadRecord = (PhdReadRecord) traceOfRead;</span>
<span class="nc" id="L136">		PhdInfo phdInfo = phdReadRecord.getPhdInfo();</span>
		
		//add read to the correct builder
<span class="nc" id="L139">		builder.addRead(read.getId(), read.getNucleotideSequence(), (int) read.getGappedStartOffset(),</span>
<span class="nc" id="L140">				read.getDirection(), read.getReadInfo().getValidRange(), phdInfo, read.getReadInfo().getUngappedFullLength());</span>
		
		//write out phd record to phdball
		try {
<span class="nc" id="L144">			phdOut.write(phdReadRecord.getPhd());</span>
<span class="nc" id="L145">		} catch (IOException e) {</span>
<span class="nc" id="L146">			throw new IllegalStateException(&quot;error writing out phd record for  &quot; + traceOfRead, e);</span>
<span class="nc" id="L147">		}</span>
<span class="nc" id="L148">		afterVisitMatch(referenceId, read,	phdReadRecord.getPhd(), phdInfo);</span>
<span class="nc" id="L149">	}</span>

	protected void afterVisitMatch(String referenceId, CasPlacedRead read,	Phd phd, PhdInfo phdInfo){
		//no-op by default
<span class="nc" id="L153">	}</span>
	
	protected void postProcess(AceContigBuilder builder){
		//no-op by default
<span class="nc" id="L157">	}</span>
	
	protected void visitAce(Range scaffoldRange, AceContig contig){
		//no-op
<span class="nc" id="L161">	}</span>
	
	@Override
	public void visitEnd() {
		//we have visited the entire cas file
		//time to write out the data
		try {
<span class="nc" id="L168">			phdOut.close();</span>
<span class="nc" id="L169">			PhdDataStore phdDataStore = new PhdFileDataStoreBuilder(phdFile)</span>
<span class="nc" id="L170">										.hint(DataStoreProviderHint.RANDOM_ACCESS_OPTIMIZE_MEMORY)</span>
<span class="nc" id="L171">										.build();</span>
<span class="nc" id="L172">			File editDir = new File(consedOutputDir, &quot;edit_dir&quot;);</span>
<span class="nc" id="L173">			File aceFile = new File(editDir, prefix + &quot;.ace.1&quot;);</span>
			//TODO customize ace writer?
<span class="nc" id="L175">			AceFileWriter aceWriter = new AceFileWriterBuilder(aceFile, phdDataStore)</span>
<span class="nc" id="L176">										.build();</span>
<span class="nc" id="L177">			Iterator&lt;Entry&lt;String, AceContigBuilder&gt;&gt; referenceEntryIter = contigBuilders.entrySet().iterator();</span>
			
			
<span class="nc bnc" id="L180" title="All 2 branches missed.">			while(referenceEntryIter.hasNext()){</span>
<span class="nc" id="L181">				Entry&lt;String, AceContigBuilder&gt; refEntry = referenceEntryIter.next();</span>
<span class="nc" id="L182">				AceContigBuilder contigBuilder = refEntry.getValue();</span>
<span class="nc" id="L183">				contigBuilder.recallConsensus(MostFrequentBasecallConsensusCaller.INSTANCE);</span>
<span class="nc" id="L184">				postProcess(contigBuilder);</span>
<span class="nc" id="L185">				visitBeginReference(refEntry.getKey());</span>
<span class="nc bnc" id="L186" title="All 2 branches missed.">				for(Entry&lt;Range,AceContig&gt; entry : ConsedUtil.split0xContig(contigBuilder,true).entrySet()){</span>
<span class="nc" id="L187">                    AceContig splitContig = entry.getValue();</span>
                    
<span class="nc" id="L189">                    visitAce(entry.getKey(), splitContig);</span>
                    
<span class="nc" id="L191">					aceWriter.write(splitContig);</span>
<span class="nc" id="L192">                }</span>
<span class="nc" id="L193">				visitEndReference();</span>
				//allows seen builders to be garbage collected if needed
<span class="nc" id="L195">				referenceEntryIter.remove();</span>
<span class="nc" id="L196">			}</span>
<span class="nc" id="L197">			aceWriter.write(new WholeAssemblyAceTag(&quot;phdBall&quot;, &quot;consed&quot;,</span>
<span class="nc" id="L198">                    DateUtil.getCurrentDate(), &quot;../phdball_dir/&quot;+phdFile.getName()));</span>
<span class="nc" id="L199">			aceWriter.close();</span>
<span class="nc" id="L200">		} catch (IOException e) {</span>
			// TODO Auto-generated catch block
<span class="nc" id="L202">			e.printStackTrace();</span>
<span class="nc" id="L203">		}</span>
		
<span class="nc" id="L205">	}</span>

	protected void visitBeginReference(String key) {
		// TODO Auto-generated method stub
		
<span class="nc" id="L210">	}</span>

	protected void visitEndReference() {
		// TODO Auto-generated method stub
		
<span class="nc" id="L215">	}</span>

	@Override
	protected StreamingIterator&lt;PhdReadRecord&gt; createFastqIterator(
			File illuminaFile) throws DataStoreException {
		try {
<span class="nc" id="L221">			FastqDataStore datastore = new FastqFileDataStoreBuilder(illuminaFile)</span>
<span class="nc" id="L222">											.hint(DataStoreProviderHint.ITERATION_ONLY)</span>
<span class="nc" id="L223">											.build();</span>
<span class="nc" id="L224">			return new FastqConsedPhdAdaptedIterator( </span>
<span class="nc" id="L225">	        		datastore.iterator(),</span>
	                illuminaFile, 
	                phdDate);
<span class="nc" id="L228">		} catch (IOException e) {</span>
<span class="nc" id="L229">			throw new IllegalStateException(&quot;fastq file no longer exists! : &quot;+ illuminaFile.getAbsolutePath());</span>
		}
	}

	@Override
	protected StreamingIterator&lt;PhdReadRecord&gt; createSffIterator(File sffFile)
			throws DataStoreException {
<span class="nc" id="L236">		return new FlowgramConsedPhdAdaptedIterator(</span>
<span class="nc" id="L237">                SffFileIterator.createNewIteratorFor(sffFile),</span>
                sffFile,
                phdDate);
	}

	@Override
	protected StreamingIterator&lt;PhdReadRecord&gt; createFastaIterator(
			File file) throws DataStoreException {
<span class="nc bnc" id="L245" title="All 2 branches missed.">		if(chromatDir ==null){</span>
			try {
<span class="nc" id="L247">				NucleotideFastaDataStore datastore = new NucleotideFastaFileDataStoreBuilder(file)</span>
<span class="nc" id="L248">																.hint(DataStoreProviderHint.ITERATION_ONLY)</span>
<span class="nc" id="L249">																.build();</span>
<span class="nc" id="L250">				return new QualFastaConsedPhdAdaptedIterator(</span>
<span class="nc" id="L251">		                datastore.iterator(),</span>
		                file,
<span class="nc" id="L253">		                phdDate, PhredQuality.valueOf(30));</span>
<span class="nc" id="L254">	        } catch (IOException e) {</span>
<span class="nc" id="L255">				throw new DataStoreException(&quot;error reading fasta file &quot;+ file.getAbsolutePath(),e);</span>
			}
		}else{
			try {
	        	//there should only be a few sanger traces so we can take the memory 
	        	//hit and store it all in memory
<span class="nc" id="L261">				NucleotideFastaDataStore datastore = new NucleotideFastaFileDataStoreBuilder(file)</span>
<span class="nc" id="L262">															.build();</span>
<span class="nc" id="L263">				return new ChromatDirFastaConsedPhdAdaptedIterator(</span>
<span class="nc" id="L264">						datastore.iterator(),</span>
						file,
<span class="nc" id="L266">		                phdDate, PhredQuality.valueOf(30),</span>
		                chromatDir);
<span class="nc" id="L268">	        } catch (IOException e) {</span>
<span class="nc" id="L269">				throw new DataStoreException(&quot;error reading fasta file for chromatogram &quot;+ file.getAbsolutePath(),e);</span>
			}
		}
	}

	
	
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.11.202310140853</span></div></body></html>