<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>CtgFileWriter.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">92_jcvi-javacommon</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.assembly.tigr.contig</a> &gt; <span class="el_source">CtgFileWriter.java</span></div><h1>CtgFileWriter.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Copyright (c) 2013 J. Craig Venter Institute.
 * 	This file is part of Jillion
 * 
 * 	 Jillion is free software: you can redistribute it and/or modify
 * 	it under the terms of the GNU General Public License as published by
 * 	the Free Software Foundation, either version 3 of the License, or
 * 	(at your option) any later version.
 * 	
 * 	 Jillion is distributed in the hope that it will be useful,
 * 	but WITHOUT ANY WARRANTY; without even the implied warranty of
 * 	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * 	GNU General Public License for more details.
 * 	
 * 	You should have received a copy of the GNU General Public License
 * 	along with  Jillion.  If not, see http://www.gnu.org/licenses
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
/*
 * Created on Mar 12, 2009
 *
 * @author dkatzel
 */
package org.jcvi.jillion.assembly.tigr.contig;

import java.io.Closeable;
import java.io.IOException;
import java.io.OutputStream;
import java.io.Serializable;
import java.io.UnsupportedEncodingException;
import java.util.Comparator;
import java.util.Set;
import java.util.TreeSet;

import org.jcvi.jillion.assembly.AssembledRead;
import org.jcvi.jillion.assembly.Contig;
import org.jcvi.jillion.core.Direction;
import org.jcvi.jillion.core.Sequence;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.residue.nt.Nucleotide;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.util.iter.StreamingIterator;
import org.jcvi.jillion.internal.core.util.JillionUtil;
/**
 * {@code CtgFileWriter} will write out {@link Contig}
 * objects in ctg format.
 * @author dkatzel
 *
 *
 */
public class CtgFileWriter implements Closeable{
<span class="nc" id="L54">    private static final CtgFormatReadSorter READ_SORTER = CtgFormatReadSorter.INSTANCE;</span>
    private final OutputStream out;
    
<span class="nc" id="L57">    public CtgFileWriter(OutputStream out) {</span>
<span class="nc" id="L58">        this.out = out;</span>
<span class="nc" id="L59">    }</span>
    public &lt;PR extends AssembledRead, C extends Contig&lt;PR&gt;&gt; void write(C contig) throws IOException,
            UnsupportedEncodingException {
<span class="nc" id="L62">        writeContigHeader(contig);</span>
<span class="nc" id="L63">        writeBases(contig.getConsensusSequence());</span>
<span class="nc" id="L64">        Set&lt;PR&gt; readsInContig = new TreeSet&lt;PR&gt;(READ_SORTER);</span>
<span class="nc" id="L65">        StreamingIterator&lt;PR&gt; iter = null;</span>
        try{
<span class="nc" id="L67">        	iter = contig.getReadIterator();</span>
<span class="nc bnc" id="L68" title="All 2 branches missed.">        	while(iter.hasNext()){</span>
<span class="nc" id="L69">        		PR placedRead = iter.next();</span>
<span class="nc" id="L70">        		readsInContig.add(placedRead);</span>
<span class="nc" id="L71">        	}</span>
        }finally{
<span class="nc" id="L73">        	IOUtil.closeAndIgnoreErrors(iter);</span>
        }
<span class="nc bnc" id="L75" title="All 2 branches missed.">        for(AssembledRead placedRead : readsInContig){</span>
<span class="nc" id="L76">            writePlacedReadHeader(placedRead, contig.getConsensusSequence());</span>
<span class="nc" id="L77">            writeBases(placedRead.getNucleotideSequence());</span>
<span class="nc" id="L78">        }</span>
<span class="nc" id="L79">    }</span>
    
   
    private void writeContigHeader(Contig&lt;? extends AssembledRead&gt; contig) throws IOException {
<span class="nc" id="L83">        String header = String.format(&quot;##%s %d %d bases&quot;,</span>
<span class="nc" id="L84">                contig.getId(), contig.getNumberOfReads(), contig.getConsensusSequence().getLength());</span>
        
<span class="nc" id="L86">        writeToOutputStream(header+&quot;\n&quot;);</span>
<span class="nc" id="L87">    }</span>

    private void writeBases(Sequence&lt;Nucleotide&gt; consensus) throws UnsupportedEncodingException, IOException {
<span class="nc" id="L90">        StringBuilder asString = new StringBuilder();</span>
<span class="nc bnc" id="L91" title="All 2 branches missed.">        for(Nucleotide glyph : consensus){</span>
<span class="nc" id="L92">            asString.append(glyph);</span>
<span class="nc" id="L93">        }</span>
        
<span class="nc" id="L95">        String result = asString.toString().replaceAll(&quot;(.{60})&quot;, &quot;$1\n&quot;);</span>
<span class="nc bnc" id="L96" title="All 2 branches missed.">        if(!result.endsWith(&quot;\n&quot;)){</span>
<span class="nc" id="L97">            result += &quot;\n&quot;;</span>
        }
<span class="nc" id="L99">        writeToOutputStream(result);</span>
        
<span class="nc" id="L101">    }</span>
    private void writeToOutputStream(final String result) throws IOException,
            UnsupportedEncodingException {
<span class="nc" id="L104">        out.write(result.getBytes(&quot;UTF-8&quot;));</span>
<span class="nc" id="L105">        out.flush();</span>
<span class="nc" id="L106">    }</span>
    
    private void writePlacedReadHeader(AssembledRead placedRead,NucleotideSequence consensus) throws IOException {
<span class="nc" id="L109">        StringBuilder header = new StringBuilder();</span>
<span class="nc" id="L110">        header.append(String.format(&quot;#%s(%d) [&quot;, placedRead.getId(), placedRead.getGappedStartOffset()));</span>
<span class="nc" id="L111">        int validLeft = (int)placedRead.getReadInfo().getValidRange().getBegin();</span>
<span class="nc" id="L112">        int validRight = (int)placedRead.getReadInfo().getValidRange().getEnd();</span>
<span class="nc bnc" id="L113" title="All 2 branches missed.">        if(placedRead.getDirection() == Direction.REVERSE){</span>
<span class="nc" id="L114">            header.append(&quot;RC&quot;);</span>
<span class="nc" id="L115">            int temp = validLeft;</span>
<span class="nc" id="L116">            validLeft = validRight;</span>
<span class="nc" id="L117">            validRight = temp;</span>
        }

<span class="nc" id="L120">        header.append(String.format(&quot;] %d bases {%d %d} &lt;%d %d&gt;&quot;,</span>
<span class="nc" id="L121">                placedRead.getNucleotideSequence().getLength(), validLeft+1, validRight+1, </span>
<span class="nc" id="L122">                placedRead.getGappedStartOffset()+1-consensus.getNumberOfGapsUntil((int) placedRead.getGappedStartOffset()), </span>
<span class="nc" id="L123">                placedRead.getGappedEndOffset()+1-consensus.getNumberOfGapsUntil((int)placedRead.getGappedEndOffset())));</span>
<span class="nc" id="L124">        header.append('\n');</span>
<span class="nc" id="L125">        writeToOutputStream(header.toString());</span>
        
<span class="nc" id="L127">    }</span>
    /**
     * {@code CtgFormatReadSorter} will sort the {@link AssembledRead}s
     * by start coordinate.  If multiple reads have the same start coordinate
     * in the contig, then those reads will be sorted by length (smallest first).  If there are still
     * multiple reads that have the same start AND the same length, then those reads
     * are sorted by their ids.
     * @author dkatzel
     *
     *
     */
<span class="nc" id="L138">    private static enum CtgFormatReadSorter implements Comparator&lt;AssembledRead&gt;, Serializable{</span>
        /**
         * Singleton instance.
         */
<span class="nc" id="L142">        INSTANCE;</span>

        /**
         * Sorts PlacedRead by offset then by read length, then by id.
         */
        @Override
        public int compare(AssembledRead o1, AssembledRead o2) {
        	
<span class="nc" id="L150">            int startComparison = JillionUtil.compare(o1.getGappedStartOffset(), o2.getGappedStartOffset());</span>
<span class="nc bnc" id="L151" title="All 2 branches missed.">            if(startComparison !=0){</span>
<span class="nc" id="L152">                return startComparison;</span>
            }
<span class="nc" id="L154">            int lengthComparison= JillionUtil.compare(o1.getGappedLength(), o2.getGappedLength());</span>
<span class="nc bnc" id="L155" title="All 2 branches missed.">            if(lengthComparison !=0){</span>
<span class="nc" id="L156">                return lengthComparison;</span>
            }
<span class="nc" id="L158">            int idLengthComparison =  JillionUtil.compare(o1.getId().length(), o2.getId().length());</span>
<span class="nc bnc" id="L159" title="All 2 branches missed.">            if(idLengthComparison !=0){</span>
<span class="nc" id="L160">                return idLengthComparison;</span>
            }
<span class="nc" id="L162">            return o1.getId().compareTo(o2.getId());</span>
        }
        
    }
    @Override
    public void close() throws IOException {
<span class="nc" id="L168">        out.close();</span>
        
<span class="nc" id="L170">    }</span>

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.11.202310140853</span></div></body></html>