<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>AceFileUtil.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">92_jcvi-javacommon</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.assembly.consed.ace</a> &gt; <span class="el_source">AceFileUtil.java</span></div><h1>AceFileUtil.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Copyright (c) 2013 J. Craig Venter Institute.
 * 	This file is part of Jillion
 * 
 * 	 Jillion is free software: you can redistribute it and/or modify
 * 	it under the terms of the GNU General Public License as published by
 * 	the Free Software Foundation, either version 3 of the License, or
 * 	(at your option) any later version.
 * 	
 * 	 Jillion is distributed in the hope that it will be useful,
 * 	but WITHOUT ANY WARRANTY; without even the implied warranty of
 * 	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * 	GNU General Public License for more details.
 * 	
 * 	You should have received a copy of the GNU General Public License
 * 	along with  Jillion.  If not, see http://www.gnu.org/licenses
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.assembly.consed.ace;

import java.io.IOException;
import java.io.OutputStream;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;

import org.jcvi.jillion.assembly.AssembledRead;
import org.jcvi.jillion.assembly.AssemblyUtil;
import org.jcvi.jillion.assembly.Contig;
import org.jcvi.jillion.assembly.consed.phd.Phd;
import org.jcvi.jillion.assembly.consed.phd.PhdDataStore;
import org.jcvi.jillion.core.Direction;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.Range.CoordinateSystem;
import org.jcvi.jillion.core.datastore.DataStoreException;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.qual.PhredQuality;
import org.jcvi.jillion.core.qual.QualitySequence;
import org.jcvi.jillion.core.qual.QualitySequenceBuilder;
import org.jcvi.jillion.core.qual.QualitySequenceDataStore;
import org.jcvi.jillion.core.residue.nt.Nucleotide;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.util.iter.StreamingIterator;
import org.jcvi.jillion.internal.core.util.JillionUtil;

/**
 * {@code AceFileUtil} is a utility class to perform 
 * common operations on ace related objects.
 * @author dkatzel
 *
 *
 */
public final class AceFileUtil {
	
    /**
     * This is the default value in consed that is used to distinguish
     * between high and low quality basecalls.  In consed
     * high quality bases are represented by uppercase
     * letters and low quality by lowercase letters.
     * Currently that value is set to 26.
     */
<span class="nc" id="L70">    public static final PhredQuality ACE_DEFAULT_HIGH_QUALITY_THRESHOLD = PhredQuality.valueOf(26);</span>
    /**
     * The date format used in consed to represent chromatogram time stamps 
     * in phd records as well as the DS lines in an ace file.
     * A read's timestamps must be identical strings in both the phd
     * and the ace for consed to make the read editable and to see the qualities
     * in the align window.
     */
<span class="nc" id="L78">    private static final DateFormat CHROMAT_DATE_TIME_FORMATTER = new SimpleDateFormat(&quot;EEE MMM d HH:mm:ss yyyy&quot;, Locale.US);</span>
    /**
     * This is the timestamp format used in some consed 
     * tags.
     */
<span class="nc" id="L83">    private static final DateFormat TAG_DATE_TIME_FORMATTER = new SimpleDateFormat(&quot;yyMMdd:HHmmss&quot;, Locale.US);</span>
    
    private static final String CONTIG_HEADER = &quot;CO %s %d %d %d %s%n&quot;;
    static{
<span class="nc" id="L87">    	CHROMAT_DATE_TIME_FORMATTER.setLenient(false);</span>
<span class="nc" id="L88">    	TAG_DATE_TIME_FORMATTER.setLenient(false);</span>
<span class="nc" id="L89">    }</span>
    
    private AceFileUtil(){
		//private constructor.
	}
    /**
     * Parse the date of a consed formatted phd date time stamp.
     * @param text the timestamp as text
     * @return the equivalent time as a {@link Date}.
     * @throws ParseException if there is a problem parsing the text.
     */
    public static synchronized Date parsePhdDate(String text) throws ParseException{
<span class="nc" id="L101">    	return CHROMAT_DATE_TIME_FORMATTER.parse(text);</span>
    }
    /**
     * format the given {@link Date} representing
     * a phd date time stamp into a consed formatted phd time stamp.
     * @param date the phd date to format.
     * @return the equivalent time as a String.
     */
    public static synchronized String formatPhdDate(Date date){
<span class="nc" id="L110">    	return CHROMAT_DATE_TIME_FORMATTER.format(date);</span>
    }
    
    /**
     * Parse the date of a consed formatted tag date time stamp.
     * @param text the timestamp as text
     * @return the equivalent time as a {@link Date}.
     * @throws ParseException if there is a problem parsing the text.
     */
    public static synchronized Date parseTagDate(String text) throws ParseException{
<span class="nc" id="L120">    	return TAG_DATE_TIME_FORMATTER.parse(text);</span>
    }
    /**
     * Format the given {@link Date} representing
     * a date time stamp for a consed tag
     * into a consed formatted time stamp.
     * @param date the date to format.
     * @return the equivalent time as a String.
     */
    public static synchronized String formatTagDate(Date date){
<span class="nc" id="L130">    	return TAG_DATE_TIME_FORMATTER.format(date);</span>
    }
    /**
     * Convert a {@link NucleotideSequence} into a string
     * where the gaps are represented by '*'s like ace files require.
     * @param basecalls input basecalls, can not be null.
     * @return a String with the all uppercase basecalls except all the '-'
     * have been replaced by '*'.
     * @throws NullPointerException if basecalls are null.
     */
    public static String convertToAcePaddedBasecalls(NucleotideSequence basecalls){
<span class="nc" id="L141">        return convertToAcePaddedBasecalls(basecalls,null);</span>
     }
    /**
     * Convert a {@link NucleotideSequence} into a string
     * where the gaps are represented by '*'s like ace files require.
     * If the optional qualities list is provided, then the returned 
     * String will return a basecalls in both upper and lowercase
     * depending on the quality value as determined by {@link AceFileUtil#ACE_DEFAULT_HIGH_QUALITY_THRESHOLD}.
     * 
     * @param basecalls input basecalls, can not be null or contain any null elements.
     * @param optionalQualities optional ungapped quality list in the same
     * orientation as the basecalls; or null if no qualities are to be used.
     * @return a String with the all uppercase basecalls except all the '-'
     * have been replaced by '*'.
     * @throws NullPointerException if basecalls are null or any Nucleotide in the basecall list 
     * is null or if the optionalQualities list is not null but contains a null in the list.
     * @throws IllegalArgumentException if optionalQualities is provided but does not have 
     * enough ungapped qualities to cover all the ungapped basecalls in the input list.
     */
     public static String convertToAcePaddedBasecalls(NucleotideSequence basecalls,QualitySequence optionalQualities){
<span class="nc" id="L161">         long length = basecalls.getLength();</span>
<span class="nc" id="L162">		StringBuilder result = new StringBuilder((int)length);</span>
<span class="nc bnc" id="L163" title="All 2 branches missed.">		Iterator&lt;PhredQuality&gt; qualityIterator = optionalQualities ==null?null : optionalQualities.iterator();</span>
<span class="nc bnc" id="L164" title="All 2 branches missed.">         for(Nucleotide base : basecalls){</span>
<span class="nc bnc" id="L165" title="All 2 branches missed.">             if(base == Nucleotide.Gap){</span>
<span class="nc" id="L166">                 result.append('*');</span>
             }
             else{
<span class="nc bnc" id="L169" title="All 2 branches missed.">                 if(optionalQualities==null){</span>
<span class="nc" id="L170">                	 result.append(base);</span>
                 }else{                 
<span class="nc bnc" id="L172" title="All 2 branches missed.">                     if(!qualityIterator.hasNext()){</span>
<span class="nc" id="L173">                         throw new IllegalArgumentException(</span>
<span class="nc" id="L174">                                 String.format(&quot;not enough ungapped qualities for input basecalls found only %d qualities&quot;,optionalQualities.getLength()));</span>
                     }
<span class="nc" id="L176">                     PhredQuality quality =qualityIterator.next();</span>
<span class="nc bnc" id="L177" title="All 2 branches missed.">                     if(quality.compareTo(ACE_DEFAULT_HIGH_QUALITY_THRESHOLD)&lt;0){</span>
<span class="nc" id="L178">                         result.append(base.toString().toLowerCase(Locale.ENGLISH));</span>
                     }
                     else{
<span class="nc" id="L181">                         result.append(base);</span>
                     }
                 }
             }
<span class="nc" id="L185">         }</span>
         
<span class="nc" id="L187">         String consedBasecalls= result.toString().replaceAll(&quot;(.{50})&quot;, &quot;$1&quot;+String.format(&quot;%n&quot;));</span>
<span class="nc bnc" id="L188" title="All 2 branches missed.">         if(length %50 ==0){</span>
             //if the last line is full, then we will have an extra %n
             //so strip it off
<span class="nc" id="L191">             return consedBasecalls.substring(0,consedBasecalls.length()-1);</span>
         }
<span class="nc" id="L193">         return consedBasecalls;</span>
     }

    
    private static String createPhdRecord(PhdInfo phdInfo){
<span class="nc" id="L198">        return String.format(&quot;DS CHROMAT_FILE: %s PHD_FILE: %s TIME: %s&quot;, </span>
                
<span class="nc" id="L200">                phdInfo.getTraceName(),</span>
<span class="nc" id="L201">                phdInfo.getPhdName(),</span>
<span class="nc" id="L202">                formatPhdDate(phdInfo.getPhdDate())</span>
                );
    }
    
    private static String createQualityRangeRecord(NucleotideSequence gappedValidBases, 
            Range ungappedValidRange, Direction dir, long ungappedFullLength){
<span class="nc" id="L208">        int numberOfGaps = gappedValidBases.getNumberOfGaps();</span>
<span class="nc" id="L209">        Range gappedValidRange =buildGappedValidRangeFor(</span>
                ungappedValidRange,numberOfGaps,dir,ungappedFullLength);

        
<span class="nc" id="L213">       return String.format(&quot;QA %d %d %d %d&quot;,</span>
<span class="nc" id="L214">                gappedValidRange.getBegin(CoordinateSystem.RESIDUE_BASED), gappedValidRange.getEnd(CoordinateSystem.RESIDUE_BASED),</span>
<span class="nc" id="L215">                gappedValidRange.getBegin(CoordinateSystem.RESIDUE_BASED), gappedValidRange.getEnd(CoordinateSystem.RESIDUE_BASED)</span>
                );
    }
    private static Range buildGappedValidRangeFor(Range ungappedValidRange, int numberOfGaps,Direction dir, long ungappedFullLength){
<span class="nc" id="L219">       Range gappedValidRange=  Range.of( </span>
<span class="nc" id="L220">               ungappedValidRange.getBegin(),</span>
<span class="nc" id="L221">               ungappedValidRange.getEnd()+numberOfGaps);</span>
        
<span class="nc bnc" id="L223" title="All 2 branches missed.">        if(dir==Direction.REVERSE){</span>
<span class="nc" id="L224">            gappedValidRange = AssemblyUtil.reverseComplementValidRange(gappedValidRange, ungappedFullLength+numberOfGaps);</span>
           
        }
<span class="nc" id="L227">        return gappedValidRange;</span>
    }
    public static synchronized String createAcePlacedReadRecord(String readId, AssembledRead placedRead, Phd phd, PhdInfo phdInfo){
        
<span class="nc" id="L231">        NucleotideSequence nucleotideSequence = placedRead.getNucleotideSequence();</span>
<span class="nc" id="L232">		final NucleotideSequence gappedValidBasecalls = nucleotideSequence; </span>
<span class="nc" id="L233">        final Range ungappedValidRange = placedRead.getReadInfo().getValidRange();</span>
<span class="nc" id="L234">        final Direction dir = placedRead.getDirection(); </span>
<span class="nc" id="L235">        final NucleotideSequence fullBasecalls = phd.getNucleotideSequence();</span>
        
<span class="nc" id="L237">        final NucleotideSequence fullGappedValidRange = AssemblyUtil.buildGappedComplementedFullRangeBases(placedRead, fullBasecalls);</span>
        final QualitySequence qualities;
<span class="nc bnc" id="L239" title="All 2 branches missed.">        if(dir ==Direction.REVERSE){</span>
<span class="nc" id="L240">        	qualities = new QualitySequenceBuilder(phd.getQualitySequence())</span>
<span class="nc" id="L241">        						.reverse()</span>
<span class="nc" id="L242">        						.build();</span>
        }else{
<span class="nc" id="L244">        	qualities = phd.getQualitySequence();</span>
        }
        
<span class="nc" id="L247">        StringBuilder readRecord = new StringBuilder();</span>
<span class="nc" id="L248">        readRecord.append(String.format(&quot;RD %s %d 0 0%n&quot;,</span>
                                            readId,
<span class="nc" id="L250">                                            fullGappedValidRange.getLength()));</span>
        
        
<span class="nc" id="L253">        readRecord.append(String.format(&quot;%s%n%n&quot;,</span>
<span class="nc" id="L254">                AceFileUtil.convertToAcePaddedBasecalls(fullGappedValidRange,qualities)));</span>
<span class="nc" id="L255">        readRecord.append(String.format(&quot;%s%n&quot;,createQualityRangeRecord(</span>
                gappedValidBasecalls,ungappedValidRange,dir, 
<span class="nc" id="L257">                fullBasecalls.getUngappedLength())));</span>
<span class="nc" id="L258">        readRecord.append(String.format(&quot;%s%n&quot;,createPhdRecord(phdInfo)));</span>
<span class="nc" id="L259">        return readRecord.toString();</span>
    }
    public static void writeAceContigHeader(String contigId, long consensusLength, long numberOfReads,
    		int numberOfBaseSegments, boolean isComplimented, OutputStream out) throws IOException{
<span class="nc" id="L263">    	writeString(String.format(CONTIG_HEADER, </span>
                contigId, 
<span class="nc" id="L265">                consensusLength,</span>
<span class="nc" id="L266">                numberOfReads,</span>
<span class="nc bnc" id="L267" title="All 2 branches missed.">                numberOfBaseSegments,</span>
                isComplimented? &quot;C&quot;:&quot;U&quot;),
                
                out);
<span class="nc" id="L271">    }</span>
    public static void writeAceFileHeader(long numberOfContigs, long numberOfReads, OutputStream out) throws IOException{
<span class="nc" id="L273">        writeString(String.format(&quot;AS %d %d%n%n&quot;, numberOfContigs, numberOfReads), out);</span>
<span class="nc" id="L274">    }</span>
    private static void writeString(String s, OutputStream out) throws IOException{
<span class="nc" id="L276">        out.write(s.getBytes(IOUtil.UTF_8));</span>
        
<span class="nc" id="L278">    }</span>
    
    
    public static void writeWholeAssemblyTag(
            WholeAssemblyAceTag wholeAssemblyTag, OutputStream out) throws IOException {
<span class="nc" id="L283">        writeString(String.format(&quot;WA{%n%s %s %s%n%s%n}%n&quot;, </span>
<span class="nc" id="L284">                wholeAssemblyTag.getType(),</span>
<span class="nc" id="L285">                wholeAssemblyTag.getCreator(),                </span>
<span class="nc" id="L286">                AceFileUtil.formatTagDate(wholeAssemblyTag.getCreationDate()),</span>
<span class="nc" id="L287">                wholeAssemblyTag.getData()), out);</span>

        
<span class="nc" id="L290">    }</span>
    public static void writeConsensusTag(ConsensusAceTag consensusTag,
            OutputStream out) throws IOException {
<span class="nc" id="L293">        StringBuilder tagBodyBuilder = new StringBuilder();</span>
<span class="nc bnc" id="L294" title="All 2 branches missed.">        if(consensusTag.getData() !=null){</span>
<span class="nc" id="L295">            tagBodyBuilder.append(consensusTag.getData());</span>
        }
<span class="nc bnc" id="L297" title="All 2 branches missed.">        if(!consensusTag.getComments().isEmpty()){</span>
<span class="nc bnc" id="L298" title="All 2 branches missed.">            for(String comment :consensusTag.getComments()){</span>
<span class="nc" id="L299">                tagBodyBuilder.append(String.format(&quot;COMMENT{%n%sC}%n&quot;,comment));            </span>
<span class="nc" id="L300">            }</span>
        }
<span class="nc" id="L302">        Range range = consensusTag.asRange();</span>
<span class="nc" id="L303">        writeString(String.format(&quot;CT{%n%s %s %s %d %d %s%s%n%s}%n&quot;, </span>
<span class="nc" id="L304">                consensusTag.getId(),</span>
<span class="nc" id="L305">                consensusTag.getType(),</span>
<span class="nc" id="L306">                consensusTag.getCreator(),</span>
<span class="nc" id="L307">                range.getBegin(),</span>
<span class="nc" id="L308">                range.getEnd(),</span>
<span class="nc" id="L309">                AceFileUtil.formatTagDate(consensusTag.getCreationDate()),</span>
<span class="nc bnc" id="L310" title="All 2 branches missed.">                consensusTag.isTransient()?&quot; NoTrans&quot;:&quot;&quot;,</span>
<span class="nc" id="L311">                        tagBodyBuilder.toString()), out);</span>
        
<span class="nc" id="L313">    }</span>
    public static void writeReadTag(ReadAceTag readTag, OutputStream out) throws IOException {
<span class="nc" id="L315">        Range range = readTag.asRange();</span>
<span class="nc" id="L316">    	writeString(String.format(&quot;RT{%n%s %s %s %d %d %s%n}%n&quot;, </span>
<span class="nc" id="L317">                        readTag.getId(),</span>
<span class="nc" id="L318">                        readTag.getType(),</span>
<span class="nc" id="L319">                        readTag.getCreator(),</span>
<span class="nc" id="L320">                        range.getBegin(),</span>
<span class="nc" id="L321">                        range.getEnd(),</span>
<span class="nc" id="L322">                        AceFileUtil.formatTagDate(readTag.getCreationDate())), out);</span>
        
<span class="nc" id="L324">    }</span>
    
    public static void writeAceContig(AceContig contig,
            PhdDataStore phdDataStore, 
            OutputStream out) throws IOException, DataStoreException{
<span class="nc" id="L329">        final NucleotideSequence consensus = contig.getConsensusSequence();</span>
<span class="nc" id="L330">        writeAceContigHeader(</span>
<span class="nc" id="L331">                contig.getId(), </span>
<span class="nc" id="L332">                consensus.getLength(),</span>
<span class="nc" id="L333">                contig.getNumberOfReads(),</span>
                0,
<span class="nc" id="L335">                contig.isComplemented(),</span>
                out);
<span class="nc" id="L337">        out.flush();</span>
<span class="nc" id="L338">        writeString(String.format(&quot;%s%n%n%n&quot;,AceFileUtil.convertToAcePaddedBasecalls(consensus)), out);</span>
<span class="nc" id="L339">        out.flush();</span>
<span class="nc" id="L340">        writeFakeUngappedConsensusQualities(consensus, out);</span>
<span class="nc" id="L341">        writeString(String.format(&quot;%n&quot;), out);</span>
<span class="nc" id="L342">        out.flush();</span>
<span class="nc" id="L343">        List&lt;IdAlignedReadInfo&gt; assembledFroms = getSortedAssembledFromsFor(contig);</span>
<span class="nc" id="L344">        StringBuilder assembledFromBuilder = new StringBuilder();</span>
<span class="nc" id="L345">        StringBuilder placedReadBuilder = new StringBuilder();</span>
        
<span class="nc bnc" id="L347" title="All 2 branches missed.">        for(IdAlignedReadInfo assembledFrom : assembledFroms){</span>
<span class="nc" id="L348">            String id = assembledFrom.getId();</span>
<span class="nc" id="L349">            final Phd phd = phdDataStore.get(id);</span>
<span class="nc" id="L350">            final AceAssembledRead realPlacedRead = contig.getRead(id);</span>
<span class="nc" id="L351">             long fullLength = realPlacedRead.getReadInfo().getUngappedFullLength();</span>
<span class="nc" id="L352">            assembledFromBuilder.append(createAssembledFromRecord(realPlacedRead,fullLength));</span>
<span class="nc" id="L353">            placedReadBuilder.append(createPlacedReadRecord(realPlacedRead, phd));</span>
<span class="nc" id="L354">        }</span>
<span class="nc" id="L355">        assembledFromBuilder.append(String.format(&quot;%n&quot;));</span>
<span class="nc" id="L356">        placedReadBuilder.append(String.format(&quot;%n&quot;));</span>
<span class="nc" id="L357">        writeString(assembledFromBuilder.toString(),out);</span>
<span class="nc" id="L358">        out.flush();</span>
<span class="nc" id="L359">        writeString(placedReadBuilder.toString(),out);</span>
<span class="nc" id="L360">        out.flush();</span>
<span class="nc" id="L361">    }</span>
    private static List&lt;IdAlignedReadInfo&gt; getSortedAssembledFromsFor(
            Contig&lt;AceAssembledRead&gt; contig){
<span class="nc" id="L364">        List&lt;IdAlignedReadInfo&gt; assembledFroms = new ArrayList&lt;IdAlignedReadInfo&gt;((int)contig.getNumberOfReads());</span>
<span class="nc" id="L365">        StreamingIterator&lt;AceAssembledRead&gt; iter = null;</span>
        try{
<span class="nc" id="L367">        	iter = contig.getReadIterator();</span>
<span class="nc bnc" id="L368" title="All 2 branches missed.">        	while(iter.hasNext()){</span>
<span class="nc" id="L369">        		AceAssembledRead read = iter.next();</span>
<span class="nc" id="L370">        		long fullLength =read.getReadInfo().getUngappedFullLength();</span>
<span class="nc" id="L371">	            assembledFroms.add(IdAlignedReadInfo.createFrom(read, fullLength));</span>
<span class="nc" id="L372">        	}</span>
        }finally{
<span class="nc" id="L374">        	IOUtil.closeAndIgnoreErrors(iter);</span>
        }
<span class="nc" id="L376">        Collections.sort(assembledFroms);</span>
<span class="nc" id="L377">        return assembledFroms;</span>
    }
    
    private static void writeFakeUngappedConsensusQualities(NucleotideSequence consensus,
            OutputStream out) throws IOException {
<span class="nc" id="L382">        StringBuilder result = new StringBuilder();</span>
<span class="nc" id="L383">        int numberOfQualitiesSoFar=0;</span>
<span class="nc bnc" id="L384" title="All 2 branches missed.">        for(int i=0; i&lt; consensus.getLength(); i++){</span>
<span class="nc" id="L385">            Nucleotide base = consensus.get(i);</span>
<span class="nc bnc" id="L386" title="All 2 branches missed.">            if(base.isGap()){</span>
<span class="nc" id="L387">                continue;</span>
            }
<span class="nc" id="L389">            result.append(&quot; 90&quot;);</span>
<span class="nc" id="L390">            numberOfQualitiesSoFar++;</span>
<span class="nc bnc" id="L391" title="All 2 branches missed.">            if(numberOfQualitiesSoFar%50==0){</span>
<span class="nc" id="L392">                result.append(String.format(&quot;%n&quot;));</span>
            }
        }
<span class="nc" id="L395">        writeString(String.format(&quot;BQ%n%s%n&quot;, result.toString()), out);</span>
<span class="nc" id="L396">    }</span>
   

    private static String createAssembledFromRecord(AceAssembledRead read, long fullLength){
<span class="nc" id="L400">    	IdAlignedReadInfo assembledFrom = IdAlignedReadInfo.createFrom(read, fullLength);</span>
<span class="nc" id="L401">        return String.format(&quot;AF %s %s %d%n&quot;,</span>
<span class="nc" id="L402">                assembledFrom.getId(),</span>
<span class="nc bnc" id="L403" title="All 2 branches missed.">                assembledFrom.getDirection()==Direction.FORWARD? &quot;U&quot;:&quot;C&quot;,</span>
<span class="nc" id="L404">                        assembledFrom.getStartOffset());</span>
    }
    
    
    private static String createPlacedReadRecord(AceAssembledRead read, Phd phd){
<span class="nc" id="L409">        return AceFileUtil.createAcePlacedReadRecord(</span>
<span class="nc" id="L410">                read.getId(),read,</span>
                phd, 
<span class="nc" id="L412">                read.getPhdInfo());</span>
        
    }
    private static final class IdAlignedReadInfo implements Comparable&lt;IdAlignedReadInfo&gt;{
    	private static final int TO_STRING_BUFFER_SIZE = 30;
		private final String id;
	    private final byte dir;
	    private final int startOffset;
<span class="nc" id="L420">	    private static final Direction[] DIRECTION_VALUES = Direction.values();</span>
	    public static IdAlignedReadInfo createFrom(AssembledRead read, long ungappedFullLength){
	        final Range validRange;
<span class="nc" id="L423">	        Direction dir = read.getDirection();</span>
<span class="nc" id="L424">	        Range readValidRange = read.getReadInfo().getValidRange();</span>
<span class="nc bnc" id="L425" title="All 2 branches missed.">	        if(dir==Direction.REVERSE){</span>
<span class="nc" id="L426">	            validRange = AssemblyUtil.reverseComplementValidRange(readValidRange, ungappedFullLength);</span>
	        }
	        else{
<span class="nc" id="L429">	            validRange = readValidRange;</span>
	        }
<span class="nc" id="L431">	        return new IdAlignedReadInfo(read.getId(), </span>
<span class="nc" id="L432">	                (int)(read.getGappedStartOffset()-validRange.getBegin()+1),dir);</span>
	    }
	    
	    
<span class="nc" id="L436">		private IdAlignedReadInfo(String id, int startOffset, Direction dir) {</span>
<span class="nc" id="L437">			this.id = id;</span>
<span class="nc" id="L438">			this.dir = (byte)dir.ordinal();</span>
<span class="nc" id="L439">			this.startOffset = startOffset;</span>
<span class="nc" id="L440">		}</span>


		@Override
	    public int hashCode() {
<span class="nc" id="L445">	        final int prime = 31;</span>
<span class="nc" id="L446">	        int result = 1;</span>
<span class="nc" id="L447">	        result = prime * result + id.hashCode();</span>
<span class="nc" id="L448">	        return result;</span>
	    }
	    @Override
	    public boolean equals(Object obj) {
<span class="nc bnc" id="L452" title="All 2 branches missed.">	        if (this == obj){</span>
<span class="nc" id="L453">	            return true;</span>
	        }
<span class="nc bnc" id="L455" title="All 2 branches missed.">	        if (obj == null){</span>
<span class="nc" id="L456">	            return false;</span>
	        }
<span class="nc bnc" id="L458" title="All 2 branches missed.">	        if (!(obj instanceof IdAlignedReadInfo)){</span>
<span class="nc" id="L459">	            return false;</span>
	        }
<span class="nc" id="L461">	        IdAlignedReadInfo other = (IdAlignedReadInfo) obj;</span>
<span class="nc" id="L462">	        return id.equals(other.getId());</span>
	    }
	    public String getId() {
<span class="nc" id="L465">	        return id;</span>
	    }

	    public int getStartOffset() {
<span class="nc" id="L469">	        return startOffset;</span>
	    }
	    
	    public Direction getDirection(){
<span class="nc" id="L473">	        return DIRECTION_VALUES[dir];</span>
	    }
	    @Override
	    public String toString() {
<span class="nc" id="L477">	        StringBuilder builder = new StringBuilder(TO_STRING_BUFFER_SIZE);</span>
<span class="nc" id="L478">	        builder.append(id).append(' ')</span>
<span class="nc" id="L479">	        		.append(startOffset)</span>
<span class="nc" id="L480">	        		.append(&quot;is complemented? &quot;)</span>
<span class="nc bnc" id="L481" title="All 2 branches missed.">	        		.append(getDirection() ==Direction.REVERSE);</span>
<span class="nc" id="L482">	        return builder.toString();</span>
	    }
	    /**
	    * Compares two AssembledFrom instances and compares them based on start offset
	    * then by Id.  This should match the order of AssembledFrom records 
	    * (and reads) in an .ace file.
	    */
	    @Override
	    public int compareTo(IdAlignedReadInfo o) {
<span class="nc" id="L491">	    	return JillionUtil.compare(getStartOffset(), o.getStartOffset());	    	</span>
	    }
    	
    }
    
    /**
     * Compute the consensus quality sequence as computed by the same algorithm consed uses.
     * @param contig the contig to compute the consensus qualities for; can not be null.
     * ConsensusQualities will only represent the qualities for the UNGAPPED 
     * consensus sequence.
     * @return a {@link QualitySequence} can not be null.
     * @throws DataStoreException 
     * @throws NullPointerException if contig is null.
     */
    public static QualitySequence computeConsensusQualities(Contig&lt;? extends AssembledRead&gt; contig, QualitySequenceDataStore readQualities) throws DataStoreException{
<span class="nc" id="L506">    	return ConsedConsensusQualityComputer.computeConsensusQualities(contig, readQualities);</span>
    }
    
    
    
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.11.202310140853</span></div></body></html>