<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>DefaultContig.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">92_jcvi-javacommon</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.internal.assembly</a> &gt; <span class="el_source">DefaultContig.java</span></div><h1>DefaultContig.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Copyright (c) 2013 J. Craig Venter Institute.
 * 	This file is part of Jillion
 * 
 * 	 Jillion is free software: you can redistribute it and/or modify
 * 	it under the terms of the GNU General Public License as published by
 * 	the Free Software Foundation, either version 3 of the License, or
 * 	(at your option) any later version.
 * 	
 * 	 Jillion is distributed in the hope that it will be useful,
 * 	but WITHOUT ANY WARRANTY; without even the implied warranty of
 * 	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * 	GNU General Public License for more details.
 * 	
 * 	You should have received a copy of the GNU General Public License
 * 	along with  Jillion.  If not, see http://www.gnu.org/licenses
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
/*
 * Created on Jan 15, 2009
 *
 * @author dkatzel
 */
package org.jcvi.jillion.internal.assembly;

import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import org.jcvi.jillion.assembly.AssembledRead;
import org.jcvi.jillion.assembly.AssembledReadBuilder;
import org.jcvi.jillion.assembly.Contig;
import org.jcvi.jillion.core.Direction;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.residue.nt.NucleotideSequenceBuilder;
import org.jcvi.jillion.core.util.MapUtil;
import org.jcvi.jillion.core.util.iter.IteratorUtil;
import org.jcvi.jillion.core.util.iter.StreamingIterator;

public final class DefaultContig&lt;T extends AssembledRead&gt; implements Contig&lt;T&gt;{

	private final NucleotideSequence consensus;
    private final String id;
    private final Map&lt;String, T&gt; mapById;
    
<span class="nc" id="L52">    public DefaultContig(String id, NucleotideSequence consensus, Set&lt;T&gt; assembledReads){</span>
<span class="nc bnc" id="L53" title="All 2 branches missed.">    	if(id==null){</span>
<span class="nc" id="L54">    		throw new NullPointerException(&quot;id can not be null&quot;);</span>
    	}
<span class="nc bnc" id="L56" title="All 2 branches missed.">    	if(consensus==null){</span>
<span class="nc" id="L57">    		throw new NullPointerException(&quot;consensus can not be null&quot;);</span>
    	}
<span class="nc" id="L59">        this.id = id;</span>
<span class="nc" id="L60">        this.consensus = consensus;</span>
<span class="nc" id="L61">        int capacity = MapUtil.computeMinHashMapSizeWithoutRehashing(assembledReads.size());</span>
<span class="nc" id="L62">        mapById = new LinkedHashMap&lt;String, T&gt;(capacity);</span>
<span class="nc bnc" id="L63" title="All 2 branches missed.">        for(T r : assembledReads){</span>
<span class="nc" id="L64">            mapById.put(r.getId(), r);</span>
<span class="nc" id="L65">        }</span>
       
<span class="nc" id="L67">    }</span>
<span class="nc" id="L68">    public DefaultContig(String id, NucleotideSequence consensus, Map&lt;String, T&gt; assembledReads){</span>
<span class="nc bnc" id="L69" title="All 2 branches missed.">    	if(id==null){</span>
<span class="nc" id="L70">    		throw new NullPointerException(&quot;id can not be null&quot;);</span>
    	}
<span class="nc bnc" id="L72" title="All 2 branches missed.">    	if(consensus==null){</span>
<span class="nc" id="L73">    		throw new NullPointerException(&quot;consensus can not be null&quot;);</span>
    	}
<span class="nc bnc" id="L75" title="All 2 branches missed.">    	if(assembledReads ==null){</span>
<span class="nc" id="L76">    		throw new NullPointerException(&quot;assembled read map can not be null&quot;);</span>
    	}
<span class="nc" id="L78">        this.id = id;</span>
<span class="nc" id="L79">        this.consensus = consensus;</span>
<span class="nc" id="L80">        mapById = assembledReads;</span>
       
<span class="nc" id="L82">    }</span>

    @Override
    public NucleotideSequence getConsensusSequence() {
<span class="nc" id="L86">        return consensus;</span>
    }

    @Override
    public String getId() {
<span class="nc" id="L91">        return id;</span>
    }

    @Override
    public long getNumberOfReads() {
<span class="nc" id="L96">        return mapById.size();</span>
    }
    @Override
    public T getRead(String id) {
<span class="nc" id="L100">        return mapById.get(id);</span>
    }
    @Override
    public StreamingIterator&lt;T&gt; getReadIterator() {       
<span class="nc" id="L104">        return IteratorUtil.createStreamingIterator(mapById.values().iterator());</span>
    }

    


    @Override
    public boolean containsRead(String placedReadId) {
<span class="nc" id="L112">        return mapById.containsKey(placedReadId);</span>
    }

	@Override
	public int hashCode() {
<span class="nc" id="L117">		final int prime = 31;</span>
<span class="nc" id="L118">		int result = 1;</span>
<span class="nc" id="L119">		result = prime * result</span>
<span class="nc" id="L120">				+ consensus.hashCode();</span>
<span class="nc" id="L121">		result = prime * result +  id.hashCode();</span>
<span class="nc" id="L122">		result = prime * result +  mapById.hashCode();</span>
<span class="nc" id="L123">		return result;</span>
	}

	@Override
	public boolean equals(Object obj) {
<span class="nc bnc" id="L128" title="All 2 branches missed.">		if (this == obj) {</span>
<span class="nc" id="L129">			return true;</span>
		}
<span class="nc bnc" id="L131" title="All 2 branches missed.">		if (obj == null) {</span>
<span class="nc" id="L132">			return false;</span>
		}
<span class="nc bnc" id="L134" title="All 2 branches missed.">		if (!(obj instanceof Contig)) {</span>
<span class="nc" id="L135">			return false;</span>
		}
<span class="nc" id="L137">		Contig&lt;?&gt; other = (Contig&lt;?&gt;) obj;</span>
		
<span class="nc bnc" id="L139" title="All 2 branches missed.">		if (!id.equals(other.getId())) {</span>
<span class="nc" id="L140">			return false;</span>
		}
<span class="nc bnc" id="L142" title="All 2 branches missed.">		if (!consensus.equals(other.getConsensusSequence())) {</span>
<span class="nc" id="L143">			return false;</span>
		}
<span class="nc bnc" id="L145" title="All 2 branches missed.">		if (getNumberOfReads()!=other.getNumberOfReads()) {</span>
<span class="nc" id="L146">			return false;</span>
		}
<span class="nc bnc" id="L148" title="All 2 branches missed.">		for(Entry&lt;String, T&gt; entry : mapById.entrySet()){</span>
<span class="nc" id="L149">			String readId = entry.getKey();</span>
<span class="nc bnc" id="L150" title="All 2 branches missed.">			if(!other.containsRead(readId)){</span>
<span class="nc" id="L151">				return false;</span>
			}
<span class="nc bnc" id="L153" title="All 2 branches missed.">			if(!entry.getValue().equals(other.getRead(readId))){</span>
<span class="nc" id="L154">				return false;</span>
			}
<span class="nc" id="L156">		}</span>
		
<span class="nc" id="L158">		return true;</span>
	}

	public static final class Builder extends AbstractContigBuilder&lt;AssembledRead, Contig&lt;AssembledRead&gt;&gt;{
        public Builder(String id, String consensus){
<span class="nc" id="L163">           this(id, new NucleotideSequenceBuilder(consensus)</span>
<span class="nc" id="L164">           					.build());</span>
<span class="nc" id="L165">        }</span>
        
        public &lt;R extends AssembledRead, C extends Contig&lt;R&gt;&gt; Builder(C copy){
<span class="nc" id="L168">            this(copy.getId(), copy.getConsensusSequence());</span>
<span class="nc" id="L169">            StreamingIterator&lt;R&gt; iter =null;</span>
            try{
<span class="nc" id="L171">            	 iter = copy.getReadIterator();</span>
<span class="nc bnc" id="L172" title="All 2 branches missed.">            	 while(iter.hasNext()){</span>
<span class="nc" id="L173">            		 R read = iter.next();</span>
<span class="nc" id="L174">            		 addRead(read);</span>
<span class="nc" id="L175">            	 }</span>
            }finally{
<span class="nc" id="L177">            	IOUtil.closeAndIgnoreErrors(iter);</span>
            }
<span class="nc" id="L179">         }</span>
        public Builder(String id, NucleotideSequence consensus){
<span class="nc" id="L181">            super(id,consensus);</span>
<span class="nc" id="L182">        }</span>
        public Builder addRead(String id, int offset,String basecalls){
<span class="nc" id="L184">            return addRead(id, offset, basecalls, Direction.FORWARD);</span>
        }
        public Builder addRead(String id, int offset,String basecalls, Direction dir){
<span class="nc" id="L187">            int numberOfGaps = computeNumberOfGapsIn(basecalls);</span>
<span class="nc" id="L188">            int ungappedLength = basecalls.length()-numberOfGaps;</span>
<span class="nc" id="L189">            return addRead(id, offset, </span>
<span class="nc" id="L190">            		Range.ofLength(ungappedLength),basecalls, </span>
            		dir,ungappedLength);
        }
        /**
         * @param basecalls
         * @return
         */
        private int computeNumberOfGapsIn(String basecalls) {
<span class="nc" id="L198">            int count=0;</span>
<span class="nc bnc" id="L199" title="All 2 branches missed.">            for(int i=0; i&lt;basecalls.length(); i++){</span>
<span class="nc bnc" id="L200" title="All 2 branches missed.">                if(basecalls.charAt(i) == '-'){</span>
<span class="nc" id="L201">                    count++;</span>
                }
            }
<span class="nc" id="L204">            return count;</span>
        }
        @Override
        public Builder addRead(String id, int offset,Range validRange, String basecalls, Direction dir, int fullUngappedLength){            
<span class="nc bnc" id="L208" title="All 2 branches missed.">            if(offset &lt;0){</span>
<span class="nc" id="L209">                throw new IllegalArgumentException(&quot;circular reads not supported&quot;);</span>
                
              }
<span class="nc" id="L212">            super.addRead(id, offset, validRange, basecalls, dir,fullUngappedLength);</span>
<span class="nc" id="L213">            return this;            </span>
        }
        
       
        public DefaultContig&lt;AssembledRead&gt; build(){
<span class="nc bnc" id="L218" title="All 2 branches missed.">        	 if(consensusCaller !=null){</span>
<span class="nc" id="L219">      			recallConsensusNow();</span>
              }
<span class="nc" id="L221">            Set&lt;AssembledRead&gt; reads = new LinkedHashSet&lt;AssembledRead&gt;();</span>
<span class="nc bnc" id="L222" title="All 2 branches missed.">            for(AssembledReadBuilder&lt;AssembledRead&gt; builder : getAllAssembledReadBuilders()){</span>
<span class="nc" id="L223">                reads.add(builder.build());</span>
<span class="nc" id="L224">            }</span>
<span class="nc" id="L225">            return new DefaultContig&lt;AssembledRead&gt;(getContigId(), getConsensusBuilder().build(), reads);</span>
        }
        /**
        * {@inheritDoc}
        */
        @Override
        protected AssembledReadBuilder&lt;AssembledRead&gt; createPlacedReadBuilder(
                AssembledRead read) {
<span class="nc" id="L233">            return DefaultAssembledRead.createBuilder(</span>
<span class="nc" id="L234">                    getConsensusBuilder().build(), </span>
<span class="nc" id="L235">                    read.getId(), </span>
<span class="nc" id="L236">                    read.getNucleotideSequence().toString(), </span>
<span class="nc" id="L237">                    (int)read.getGappedStartOffset(), </span>
<span class="nc" id="L238">                    read.getDirection(), </span>
<span class="nc" id="L239">                    read.getReadInfo().getValidRange(),</span>
<span class="nc" id="L240">                    read.getReadInfo().getUngappedFullLength());</span>
        }
        /**
        * {@inheritDoc}
        */
        @Override
        protected AssembledReadBuilder&lt;AssembledRead&gt; createPlacedReadBuilder(
                String id, int offset, Range validRange, String basecalls,
                Direction dir, int fullUngappedLength) {
<span class="nc" id="L249">            return DefaultAssembledRead.createBuilder(</span>
<span class="nc" id="L250">                    getConsensusBuilder().build(), </span>
                    id, 
                    basecalls, 
                    offset, 
                    dir, 
                    validRange,
                    fullUngappedLength);
        }
    }

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.11.202310140853</span></div></body></html>