<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>FastqUtil.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">92_jcvi-javacommon</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.trace.fastq</a> &gt; <span class="el_source">FastqUtil.java</span></div><h1>FastqUtil.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Copyright (c) 2013 J. Craig Venter Institute.
 * 	This file is part of Jillion
 * 
 * 	 Jillion is free software: you can redistribute it and/or modify
 * 	it under the terms of the GNU General Public License as published by
 * 	the Free Software Foundation, either version 3 of the License, or
 * 	(at your option) any later version.
 * 	
 * 	 Jillion is distributed in the hope that it will be useful,
 * 	but WITHOUT ANY WARRANTY; without even the implied warranty of
 * 	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * 	GNU General Public License for more details.
 * 	
 * 	You should have received a copy of the GNU General Public License
 * 	along with  Jillion.  If not, see http://www.gnu.org/licenses
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.trace.fastq;

import java.io.File;
import java.io.IOException;
import java.util.EnumMap;
import java.util.Map;
import java.util.SortedMap;
import java.util.regex.Pattern;

import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.util.MapValueComparator;

/**
 * {@code FastqUtil} is a utility class for working with 
 * FASTQ data.
 * @author dkatzel
 *
 *
 */
public final class FastqUtil {


    /**
     * This is the {@link Pattern} to parse
     * the sequence record defline of a FASTQ record.
     * Group 1 will be the read id
     * Group 3 will be the optional comment if there is one,
     * or null if there isn't a comment.
     */
<span class="nc" id="L50">    public static final Pattern SEQ_DEFLINE_PATTERN = Pattern.compile(&quot;^@(\\S+)(\\s+)?(.+$)?&quot;);</span>
    /**
     * This is the {@link Pattern} to parse
     * the quality record defline of a FASTQ record.
     * Group 1 will be the optional id of the read if there is one
     * or null if there isn't an id.  If the id exists,
     * then it should match the id of the seq defline.
     */
<span class="nc" id="L58">    public static final Pattern QUAL_DEFLINE_PATTERN = Pattern.compile(&quot;^\\+(.+$)?&quot;);</span>
    

	private FastqUtil(){
		//can not instantiate
	}
    
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the qualities in the given fastq file. 
     * This method works by analyzing the encoded quality value ranges for each
     * read getting inspected.  Since all the different quality encodings have
     * overlapping values the {@link FastqQualityCodec} used for the entire file
     * can not always be determined from just analyzing a few reads.
     * Therefore, this method might have to parse
     * ALL records in the fastq file to guarantee that the
     * correct {@link FastqQualityCodec} is returned.
     * @param fastqFile a fastq File, must exist and can not be null.
     * @return an instance of {@link FastqQualityCodec} which is the one
     * that is most likely able to decode the quality values for all the reads
     * in the file (will never be null).
     * @throws IOException if there is a problem reading the file.
     * @throws NullPointerException if fastqFile is null.
     * @throws IllegalArgumentException if the given quality string is empty
     * or if it contains any characters out of range of any known
     * quality encoding formats.
     * @throws IllegalStateException if the fastq file does not contain 
     * any records.
     */
    public static FastqQualityCodec guessQualityCodecUsed(File fastqFile) throws IOException{
<span class="nc" id="L88">    	return guessQualityCodecUsed(fastqFile, Integer.MAX_VALUE);</span>
    }
    
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the qualities in the given fastq file. 
     * This method works by analyzing the encoded quality value ranges for each
     * read getting inspected.  Since all the different quality encodings have
     * overlapping values the {@link FastqQualityCodec} used for the entire file
     * can not always be determined from just analyzing a few reads.
     * Therefore, this method must
     * parse many (probably hundreds) of reads in the fastq file in order to get
     * a large enough sampling size to confidently return a {@link FastqQualityCodec}.
     * This method will keep looking at reads in the fastq file until
     * either the file ends, or it has looked at {@code numberOfReadsToInspect}.
     * @param fastqFile a fastq File, must exist and can not be null.
     * @param numberOfReadsToInspect the number of reads in the file to analyze.
     * 
     * @return an instance of {@link FastqQualityCodec} which is the one
     * that is most likely able to decode the quality values for all the reads
     * in the file (will never be null).
     * @throws IOException if there is a problem reading the file.
     * @throws NullPointerException if fastqFile is null.
     * @throws IllegalArgumentException if the given quality string is empty
     * or if it contains any characters out of range of any known
     * quality encoding formats.
     * @throws IllegalArgumentException if numberOfReadsToInspect is &lt; 1.
     * @throws IllegalStateException if the fastq file does not contain 
     * any records.
     */
    public static FastqQualityCodec guessQualityCodecUsed(File fastqFile, int numberOfReadsToInspect) throws IOException{
<span class="nc bnc" id="L119" title="All 2 branches missed.">    	if(numberOfReadsToInspect &lt;1){</span>
<span class="nc" id="L120">    		throw new IllegalArgumentException(&quot;number of reads to inspect must be &gt;=1&quot;);</span>
    	}
<span class="nc" id="L122">    	FastqQualityCodecDetectorVisitor detectorVisitor =new FastqQualityCodecDetectorVisitor(numberOfReadsToInspect);</span>
<span class="nc" id="L123">    	FastqFileParser.create(fastqFile).accept(detectorVisitor);</span>
<span class="nc" id="L124">    	return detectorVisitor.getDetectedCodec();</span>
    }
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the given qualities.
     * @param encodedQualities a String of fastq encoded qualities
     * @return an instance of {@link FastqQualityCodec} which could have been 
     * used to encode the given qualities (will never be null).
     * @throws NullPointerException if the given qualities are null.
     * @throws IllegalArgumentException if the given quality string is empty
     * or if it contains any characters out of range of any known
     * quality encoding formats.
     */
    static FastqQualityCodec guessQualityCodecUsed(String encodedQualities){
<span class="nc bnc" id="L138" title="All 2 branches missed.">    	if(encodedQualities.isEmpty()){</span>
<span class="nc" id="L139">    		throw new IllegalArgumentException(&quot;encoded qualities can not be empty&quot;);</span>
    	}
    	//sanger uses 33 as an offset so any ascii values around there will 
    	//automatically be sanger
    	
    	//solexa and illumina encoding have offsets of 64 so they look very similar
    	//except solexa uses a different log scale and can actually have scores as low
    	//as -5 (ascii value 59) so any values from ascii 59-63 mean solexa
<span class="nc" id="L147">    	boolean hasSolexaOnlyValues=false;</span>
<span class="nc" id="L148">    	int maxQuality=Integer.MIN_VALUE;</span>
<span class="nc" id="L149">    	int minQuality = Integer.MAX_VALUE;</span>
    	
<span class="nc bnc" id="L151" title="All 2 branches missed.">    	for(int i=0; i&lt;encodedQualities.length(); i++){</span>
<span class="nc" id="L152">    		int asciiValue = encodedQualities.charAt(i);</span>
<span class="nc bnc" id="L153" title="All 2 branches missed.">    		if(asciiValue &lt;33){</span>
<span class="nc" id="L154">    			throw new IllegalArgumentException(</span>
<span class="nc" id="L155">    					String.format(</span>
    							&quot;invalid encoded qualities has out of range ascii value %d : '%s'&quot;, 
<span class="nc" id="L157">    							asciiValue,</span>
    							encodedQualities));
    		}
<span class="nc bnc" id="L160" title="All 2 branches missed.">    		if(asciiValue &lt;59){</span>
<span class="nc" id="L161">    			return FastqQualityCodec.SANGER;</span>
    		}
<span class="nc bnc" id="L163" title="All 2 branches missed.">    		if(asciiValue &lt; 64){</span>
<span class="nc" id="L164">    			hasSolexaOnlyValues=true;</span>
    		}
<span class="nc bnc" id="L166" title="All 2 branches missed.">    		if(asciiValue &gt; maxQuality){</span>
<span class="nc" id="L167">    			maxQuality = asciiValue;</span>
    		}
<span class="nc bnc" id="L169" title="All 2 branches missed.">    		if(asciiValue &lt; minQuality){</span>
<span class="nc" id="L170">    			minQuality = asciiValue;</span>
    		}
    	}
<span class="nc bnc" id="L173" title="All 2 branches missed.">    	if(hasSolexaOnlyValues){</span>
<span class="nc" id="L174">    		return FastqQualityCodec.SOLEXA;</span>
    	}
    	//if we get here then we only saw encoded values from ascii 64 +
    	//assume illumina, solexa scaling is so close at good quality anyway
    	//that I don't think it matters.
    	
<span class="nc" id="L180">    	return FastqQualityCodec.ILLUMINA;</span>
    }
    /**
     * Inner class that visits the first X reads of a fastq file
     * and notes the {@link FastqQualityCodec} that can be used.
     * After a sample size of X reads is analyzed, the most frequent
     * codec is returned.
     * @author dkatzel
     *
     */
    private static final class FastqQualityCodecDetectorVisitor implements FastqVisitor{
<span class="nc" id="L191">    	private int numberOfRecordsVisited=0;</span>
    	private final int maxNumberOfRecordsToVisit;
<span class="nc" id="L193">    	private FastqQualityCodec detectedCodec=null;</span>
		
<span class="nc" id="L195">    	FastqQualityCodecDetectorRecordVisitor recordVisitor = new FastqQualityCodecDetectorRecordVisitor();</span>
    	
<span class="nc" id="L197">		public FastqQualityCodecDetectorVisitor(int maxNumberOfRecordsToVisit) {</span>
<span class="nc" id="L198">			this.maxNumberOfRecordsToVisit = maxNumberOfRecordsToVisit;</span>
<span class="nc" id="L199">		}</span>
		public FastqQualityCodec getDetectedCodec() {
<span class="nc" id="L201">			return detectedCodec;</span>
		}
		@Override
		public FastqRecordVisitor visitDefline(FastqVisitorCallback callback,
				String id, String optionalComment) {
<span class="nc bnc" id="L206" title="All 2 branches missed.">			if(numberOfRecordsVisited&lt;maxNumberOfRecordsToVisit){</span>
<span class="nc" id="L207">				numberOfRecordsVisited++;</span>
<span class="nc" id="L208">				return recordVisitor;</span>
			}
<span class="nc" id="L210">			callback.haltParsing();</span>
<span class="nc" id="L211">			return null;</span>
		}
		/**
		 * This method will be the most frequently
		 * one called since most fastq files
		 * are huge so {@link org.jcvi.jillion.trace.fastq.FastqVisitor.FastqVisitorCallback#haltParsing()} will get called
		 * 
		 */
		@Override
		public void halted() {
<span class="nc" id="L221">			detectedCodec= recordVisitor.getMostFrequentCodec();			</span>
<span class="nc" id="L222">		}</span>
		/**
		 * It is possible that the fastq
		 * file is so small that 
		 * {@link org.jcvi.jillion.trace.fastq.FastqVisitor.FastqVisitorCallback#haltParsing()}
		 * will not get called.
		 * 
		 */
		@Override
		public void visitEnd() {			
<span class="nc" id="L232">	        detectedCodec= recordVisitor.getMostFrequentCodec();			</span>
<span class="nc" id="L233">		}</span>
    	
    }
    
<span class="nc" id="L237">    private static final class FastqQualityCodecDetectorRecordVisitor implements FastqRecordVisitor{</span>
<span class="nc" id="L238">    	private int numSanger=0;</span>
<span class="nc" id="L239">    	private int numIllumina=0;</span>
<span class="nc" id="L240">    	private int numSolexa=0;</span>
    	
		@Override
		public void visitNucleotides(NucleotideSequence nucleotides) {
			//no-op			
<span class="nc" id="L245">		}</span>

		@Override
		public void visitEncodedQualities(String encodedQualities) {
<span class="nc" id="L249">			FastqQualityCodec codec =FastqUtil.guessQualityCodecUsed(encodedQualities);</span>
<span class="nc bnc" id="L250" title="All 4 branches missed.">            switch(codec){</span>
<span class="nc" id="L251">                case SANGER : numSanger++;</span>
<span class="nc" id="L252">                				break;</span>
<span class="nc" id="L253">                case ILLUMINA : numIllumina++;</span>
<span class="nc" id="L254">								break;</span>
<span class="nc" id="L255">                case SOLEXA : numSolexa++;</span>
<span class="nc" id="L256">								break;</span>
				default:
<span class="nc" id="L258">					throw new IllegalArgumentException(&quot;unknown codec: &quot;+ codec);</span>
            }
			
<span class="nc" id="L261">		}</span>

		@Override
		public void visitEnd() {
			//no-op			
<span class="nc" id="L266">		}</span>
    	
		@Override
		public void halted() {
			//no-op			
<span class="nc" id="L271">		}</span>
		
		public FastqQualityCodec getMostFrequentCodec(){
<span class="nc bnc" id="L274" title="All 6 branches missed.">			if(numSanger==0 &amp;&amp; numSolexa==0 &amp;&amp; numIllumina==0){</span>
<span class="nc" id="L275">				throw new IllegalStateException(&quot;fastq file must not be empty&quot;);</span>
			}
<span class="nc" id="L277">			Map&lt;FastqQualityCodec, Integer&gt; map = new EnumMap&lt;FastqQualityCodec, Integer&gt;(FastqQualityCodec.class);</span>
<span class="nc" id="L278">	        map.put(FastqQualityCodec.SANGER, Integer.valueOf(numSanger));</span>
<span class="nc" id="L279">	        map.put(FastqQualityCodec.SOLEXA, Integer.valueOf(numSolexa));</span>
<span class="nc" id="L280">	        map.put(FastqQualityCodec.ILLUMINA, Integer.valueOf(numIllumina));</span>
<span class="nc" id="L281">	        SortedMap&lt;FastqQualityCodec, Integer&gt; sortedMap = MapValueComparator.sortDescending(map);</span>
<span class="nc" id="L282">	        return sortedMap.firstKey();</span>
		}
    }
    
    
   
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.11.202310140853</span></div></body></html>